src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+{
+	/* determine the number of TMUs involved */
+	UINT32 tmus = 0, texmode0 = 0, texmode1 = 0;
+	if (!FBIINIT3_DISABLE_TMUS(v->reg[fbiInit3].u) && FBZCP_TEXTURE_ENABLE(v->reg[fbzColorPath].u))
+	{
+		tmus = 1;
+		texmode0 = v->tmu[0].reg[textureMode].u;
+		if (v->chipmask & 0x04)
+		{
+			tmus = 2;
+			texmode1 = v->tmu[1].reg[textureMode].u;
+		}
+		if (tworker.disable_bilinear_filter) //force disable bilinear filter
+		{
+			texmode0 &= ~6;
+			texmode1 &= ~6;
+		}
+	}
+
+	/* compute the slopes for each portion of the triangle */
+	poly_vertex v1 = tworker.v1, v2 = tworker.v2, v3 = tworker.v3;
+	float dxdy_v1v2 = (v2.y == v1.y) ? 0.0f : (v2.x - v1.x) / (v2.y - v1.y);
+	float dxdy_v1v3 = (v3.y == v1.y) ? 0.0f : (v3.x - v1.x) / (v3.y - v1.y);
+	float dxdy_v2v3 = (v3.y == v2.y) ? 0.0f : (v3.x - v2.x) / (v3.y - v2.y);
+
+	stats_block my_stats;
+	INT32 from = tworker.totalpix * worktstart / TRIANGLE_WORKERS;
+	INT32 to   = tworker.totalpix * worktend   / TRIANGLE_WORKERS;
+	for (INT32 curscan = tworker.v1y, scanend = tworker.v3y, sumpix = 0, lastsum = 0; curscan != scanend && lastsum < to; lastsum = sumpix, curscan++)
+	{
+		float fully = (float)(curscan) + 0.5f;
+		float startx = v1.x + (fully - v1.y) * dxdy_v1v3;
+
+		/* compute the ending X based on which part of the triangle we're in */
+		float stopx = (fully < v2.y ? (v1.x + (fully - v1.y) * dxdy_v1v2) : (v2.x + (fully - v2.y) * dxdy_v2v3));
+
+		/* clamp to full pixels */
+		poly_extent extent;
+		extent.startx = round_coordinate(startx);
+		extent.stopx = round_coordinate(stopx);
+
+		/* force start < stop */
+		if (extent.startx >= extent.stopx)
+		{
+			if (extent.startx == extent.stopx) continue;
+			std::swap(extent.startx, extent.stopx);
+		}
+
+		sumpix += (extent.stopx - extent.startx);
+
+		if (sumpix <= from)
+			continue;
+		if (lastsum < from)
+			extent.startx += (from - lastsum);
+		if (sumpix > to)
+			extent.stopx -= (sumpix - to);
+
+		raster_generic(v, tmus, texmode0, texmode1, tworker.drawbuf, curscan, &extent, my_stats);
+	}
+	sum_statistics(&v->thread_stats[worktstart], &my_stats);
+}
+
+static int triangle_worker_thread_func(void* p)
+{
+	triangle_worker& tworker = v->tworker;
+	for (INT32 tnum = (INT32)(size_t)p; tworker.threads_active;)
+	{
+		SDL_SemWait(tworker.sembegin[tnum]);
+		if (tworker.threads_active)
+			triangle_worker_work(tworker, tnum, tnum + 1);
+		tworker.done[tnum] = true;
+	}
+	return 0;
+}
+
+static void triangle_worker_shutdown(triangle_worker& tworker)
+{
+	if (!tworker.threads_active) return;
+	tworker.threads_active = false;
+	for (size_t i = 0; i != TRIANGLE_THREADS; i++) tworker.done[i] = false;
+	for (size_t i = 0; i != TRIANGLE_THREADS; i++) SDL_SemPost(tworker.sembegin[i]);
+	recheckdone:
+	for (size_t i = 0; i != TRIANGLE_THREADS; i++) if (!tworker.done[i]) goto recheckdone;
+	for (size_t i = 0; i != TRIANGLE_THREADS; i++) SDL_DestroySemaphore(tworker.sembegin[i]);
+}
```

Comment:
This thread-shutdown function should cleanup the zombie threads using  [SDL_WaitThread](https://wiki.libsdl.org/SDL2/SDL_WaitThread) because they're not detached. 
> "Threads that haven't been detached will remain (as a "zombie") until this function cleans them up. Not doing so is a resource leak."

> "Once a thread has been cleaned up through this function, the [SDL_Thread](https://wiki.libsdl.org/SDL2/SDL_Thread) that references it becomes invalid and should not be referenced again. As such, only one thread may call [SDL_WaitThread](https://wiki.libsdl.org/SDL2/SDL_WaitThread)() on another."

Reply:
To be addressed, probably while investigating the desired number of worker threads.